stl/src/fiopen.cpp

```diff
@@ -58,13 +58,15 @@ FILE* _Xfiopen(const CharT* filename, ios_base::openmode mode, int prot) {
 
     mode &= ~(ios_base::ate | ios_base::_Nocreate | ios_base::_Noreplace);
 
-    int n = 0;
-    while (valid[n] != 0 && valid[n] != mode) { // look for a valid mode
-        ++n;
+    if (mode == 0) {
+        return nullptr; // no valid mode
     }
 
-    if (valid[n] == 0) {
-        return nullptr; // no valid mode
+    int n = 0;
+    for (; valid[n] != mode; ++n) { // look for a valid mode
+        if (valid[n] == 0) {
+            return nullptr; // no valid mode
+        }

```

Comment:
This doesn't strike me as an improvement: we started with two branches poor loop iteration, and still have two branches per loop iteration, and the trivially-optimized-away branch after the loop has become a very hard to optimize away branch before the loop. I think we could improve a bit by abandoning standard loop structure:
```c++
int n = 0;
for (;; ++n) {
  if (valid[n] == 0) {
    return nullptr;
  }
  if (valid[n] == mode) {
    break;
  }
}
```
which coalesces the before-or-after the loop check into the per-iteration checks (which is the form a good optimizer likely transforms the original code into anyway).

Reply:
Is the goal performance? It's odd to make instruction level optimizations for file i/o (which is ultimately a syscall), and it is odd to favor negative case in such optimization

Reply:
Negative case yes, but at the same time, this check is always done anyway. If it wasn’t, then you’d be right. 

If checks are inevitable, put them first so in those cases we can save time